src/key_gen/mod.rs

```diff
@@ -530,17 +541,27 @@ impl<S: SecretId> KeyGen<S> {
                     }
                 }
             }
-            Err(_fault) => {
+            Err(fault) => {
                 let msg = Message::<S::PublicId>::Acknowledgment {
                     key_gen_id: sender_index,
                     ack,
                 };
-                let invalid_ack = serialize(&msg)?;
-                self.pending_complain_messages.push(Message::Complaint {
-                    key_gen_id: self.our_index,
-                    target: sender_index,
-                    msg: invalid_ack,
-                });
+                debug!(
+                    "{:?} complain {:?} with Error {:?}",
+                    self, sender_index, fault
+                );
+                if fault == AcknowledgmentFault::MissingPart {
+                    // TODO: consider carry out caching internally.
+                    debug!("{:?} MissingPart on Ack not causing a complain, return with error to trigger an outside cache", self);
+                    return Err(Error::MissingPart);
+                } else {
```

Comment:
It would be cleaner to take this part out into its own `match` branch:
```rust
match self.handle_ack_or_fault(...) {
    Ok(()) => ...
    Err(AcknowledgementFault::MissingPart) => {
        // ...
        return Err(Error::MissingPart);
    }
    Err(fault) => {
        // the other thing here
    }
}
```